p2p/discover/v4_udp.go

```diff
@@ -475,16 +513,53 @@ func (t *UDPv4) findnode(toid enode.ID, toaddr *net.UDPAddr, target encPubkey) (
 		}
 		return true, nreceived >= bucketSize
 	})
-	t.send(toaddr, toid, p_findnodeV4, &findnodeV4{
+	t.send(toaddr, toid, &findnodeV4{
 		Target:     target,
 		Expiration: uint64(time.Now().Add(expiration).Unix()),
 	})
-	return nodes, <-errc
+	return nodes, <-rm.errc
+}
+
+// requestENR sends enrRequest to the given node and waits for a response.
+func (t *UDPv4) requestENR(n *enode.Node) (*enode.Node, error) {
+	addr := &net.UDPAddr{IP: n.IP(), Port: n.UDP()}
+	t.ensureBond(n.ID(), addr)
+
+	req := &enrRequestV4{
+		Expiration: uint64(time.Now().Add(expiration).Unix()),
+	}
+	packet, hash, err := t.encode(t.priv, req)
+	if err != nil {
+		return nil, err
+	}
+	// Add a matcher for the reply to the pending reply queue. Responses are matched if
+	// they reference the request we're about to send.
+	rm := t.pending(n.ID(), addr.IP, p_enrResponseV4, func(r interface{}) (matched bool, requestDone bool) {
+		matched = bytes.Equal(r.(*enrResponseV4).ReplyTok, hash)
+		return matched, matched
+	})
+	// Send the packet and wait for the reply.
+	t.write(addr, n.ID(), req.name(), packet)
+	if err := <-rm.errc; err != nil {
+		return nil, err
+	}
+	// Verify the response record.
+	respN, err := enode.New(enode.ValidSchemes, &rm.reply.(*enrResponseV4).Record)
+	if err != nil {
+		return nil, err
+	}
+	if respN.ID() != n.ID() {
+		return nil, fmt.Errorf("invalid ID in response record")
+	}
+	if err := netutil.CheckRelayIP(addr.IP, respN.IP()); err != nil {
```

Comment:
As an attacker at this point, can I 
a) supply many (millions?) ENR records all with the same IP (me) to populate the DHT with spam? 
b) Then can I switch the IP addresses to focus traffic on some node?
Both would be attacks.

Reply:
No, these kinds of attacks are not possible. The response can only contain a single record and the regular IP limits on the table still apply even for record updates.